src/parquet/schema.cc

```diff
@@ -620,6 +620,15 @@ const ColumnDescriptor* SchemaDescriptor::Column(int i) const {
   return &leaves_[i];
 }
 
+int SchemaDescriptor::ColumnIndex(const NodePtr& node) const {
```

Comment:
I agree that adding the classes will add unnecessary code complexity. 
I will make the changes for the ToDotString method. I'm just think that about caching, since `NodePtr`s can be moved around (when manipulating schemas for example), it can be a bit tricky and the PR is kind of big as it is. 
So I offer that for the time being I won't add caching but leave a TODO comment for it, since it seems to me that in any case this is not called a lot and not worth the effort, at least currently.